idx/memory/memory.go

```diff
@@ -363,6 +367,94 @@ func (m *UnpartitionedMemoryIdx) UpdateArchive(archive idx.Archive) {
 	*(m.defById[archive.Id]) = archive
 }
 
+// MetaTagRecordUpsert inserts or updates a meta record, depending on whether
+// it already exists or is new. The identity of a record is determined by its
+// queries, if the set of queries in the given record already exists in another
+// record, then the existing record will be updated, otherwise a new one gets
+// created.
+// The return values are:
+// 1) The relevant meta record as it is after this operation
+// 2) A bool that is true if the record has been created, or false if updated
+// 3) An error which is nil if no error has occurred
+func (m *UnpartitionedMemoryIdx) MetaTagRecordUpsert(orgId uint32, rawRecord idx.MetaTagRecord) (idx.MetaTagRecord, bool, error) {
+	res := idx.MetaTagRecord{}
+
+	if !TagSupport {
+		log.Warn("memory-idx: received tag query, but tag support is disabled")
+		return res, false, fmt.Errorf("Tag support is disabled")
+	}
+
+	var mtr metaTagRecords
+	var mti metaTagIndex
+	var ok bool
+
+	m.Lock()
+	defer m.Unlock()
+
+	if mtr, ok = m.metaTagRecords[orgId]; !ok {
+		mtr = make(metaTagRecords)
+		m.metaTagRecords[orgId] = mtr
+	}
+
+	if mti, ok = m.metaTags[orgId]; !ok {
+		mti = make(metaTagIndex)
+		m.metaTags[orgId] = mti
+	}
+
+	hash, record, oldHash, oldRecord, err := mtr.upsert(rawRecord.MetaTags, rawRecord.Queries)
+	if err != nil {
+		return res, false, err
+	}
+
+	builder := strings.Builder{}
+	res.ID = hash
+	res.MetaTags = record.metaTagStrings(&builder)
+	res.Queries = record.queryStrings(&builder)
+
+	// if this upsert has updated a previously existing record, then we remove its entries
+	// from the metaTagIndex before inserting the new ones
+	if oldRecord != nil {
+		for _, keyValue := range oldRecord.metaTags {
+			mti.deleteRecord(keyValue, oldHash)
+		}
+
+		for _, keyValue := range record.metaTags {
+			mti.insertRecord(keyValue, hash)
+		}
+
+		return res, false, nil
+	} else {
+		for _, keyValue := range record.metaTags {
+			mti.insertRecord(keyValue, hash)
+		}
+
+		return res, true, nil
+	}
+}
+
+func (m *UnpartitionedMemoryIdx) MetaTagRecordList(orgId uint32) []idx.MetaTagRecord {
+	builder := strings.Builder{}
+	res := make([]idx.MetaTagRecord, 0, len(m.metaTagRecords))
```

Comment:
```suggestion
	var res []idx.MetaTagRecord
```
`len(m.metaTagRecords)` is just however many orgs we've seen and is irelevant